api_app/analyzers_manager/observable_analyzers/greynoise_labs.py

```diff
@@ -0,0 +1,144 @@
+                        raise AnalyzerRunException(f"Failed extraction from {key} db")
+                    if not os.path.exists(value["db_location"]):
+                        raise AnalyzerRunException(
+                            f"database location {value['db_location']} does not exist"
+                        )
+
+                    with open(value["db_location"], "r", encoding="utf-8") as f:
+                        db = f.read()
+
+                    db_list = db.split("\n")
+                    if self.observable_name in db_list:
+                        result[key] = {"found": True}
+                    else:
+                        result[key] = {"found": False}
+                    continue
+
+                json_body = {
+                    "query": value["query_string"],
+                    "variables": {"ip": f"{self.observable_name}"},
+                }
+                response = requests.post(headers=headers, json=json_body, url=url)
+                response.raise_for_status()
+                result[key] = response.json()
+        except requests.RequestException as e:
+            raise AnalyzerRunException(e)
+
+        return result
+
+    @classmethod
+    def _get_auth_token(cls):
+        for plugin in PluginConfig.objects.filter(
+            parameter__python_module=cls.python_module,
+            parameter__is_secret=True,
+            parameter__name="auth_token",
+        ):
+            if plugin.value:
+                return plugin.value
+        return None
+
+    @classmethod
+    def _update_db(cls, auth_token: str):
+        headers = {
+            "Content-Type": "application/json",
+            "Authorization": f"Bearer {auth_token}",
+        }
+
+        try:
+            logger.info("Fetching data from greynoise API.....")
+            response = requests.post(
+                headers=headers,
+                json={"query": queries["topc2s"]["query_string"]},
+                url=url,
+            )
+            response.raise_for_status()
+            topc2s_data = response.json()
+
+            with open(db_location, "w", encoding="utf-8") as f:
+                for value in topc2s_data["data"]["topC2s"]["c2s"]:
+                    ip = value["source_ip"]
+                    if ip:
+                        f.write(f"{ip}\n")
+
+            if not os.path.exists(db_location):
+                return False
+
+            logger.info("Data fetched from greynoise API.....")
+            return True
+        except Exception as e:
+            logger.exception(e)
+
+    @classmethod
+    def update(cls):
+        auth_token = cls._get_auth_token()
+        if auth_token:
+            return cls._update_db(auth_token=auth_token)
+        return False
+
+    @classmethod
+    def _monkeypatch(cls):
+        patches = [
+            if_mock_connections(
+                patch("requests.post", return_value=MockUpResponse({}, 200))
```

Comment:
can you please write examples of outputs here in the tests? like the ones that you shared with me

In this way tests would run with a real output and we could also save an example of their reports here.

Please do a mock for every requests you do (2). See `feodo tracker` as an example

Reply:
Like one request whose ip is in noiseRank and one which is not in noiseRank, right? @mlodic

Reply:
basically here the tests will cycle through the list of mocks that you write and use one of them every time the analyzer would try to do a http request of the chosen method.

So basically one mock for each request that you make, so one for each endpoint in greynoise